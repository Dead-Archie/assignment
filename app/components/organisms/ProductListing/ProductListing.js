import React from 'react';
import PropTypes from 'prop-types';
import Router from 'next/router';
import ProductListGrid from '../ProductListGrid';
import { queryParamStr } from '../../../utils/utils';
import withStyles from '../../../lib/withStyles';
import styles from './ProductListing.style';

// eslint-disable-next-line react/prefer-stateless-function
class ProductListing extends React.Component {
  constructor(props) {
    super(props);
    this.searchRef = React.createRef();
    this.sortRef = React.createRef();
    this.state = {
      productList: props.data,
    };
  }

  componentDidUpdate(prevProps, prevState) {
    const { data: currentData } = this.props;
    const { productList } = prevState;
    const { value } = this.sortRef.current;

    if (productList !== currentData && currentData) {
      if (value === 'dsnd') {
        // eslint-disable-next-line react/no-did-update-set-state
        this.setState({
          productList: currentData.sort((a, b) => b.id - a.id),
        });
      } else {
        // eslint-disable-next-line react/no-did-update-set-state
        this.setState({
          productList: currentData.sort((a, b) => a.id - b.id),
        });
      }
    }
  }

  removePils = name => {
    const { query } = this.props;
    delete query[name];
    const queryStr = queryParamStr(query);

    Router.push(`/dashboard?${queryStr}`);
  };

  filterPils = name => {
    if (name === 'name') {
      return false;
    }
    const { query } = this.props;
    const filterPill = query[name];
    return (
      <button className="filter-pills" onClick={() => this.removePils(name)}>
        {filterPill}
      </button>
    );
  };

  getSortedValue = () => {
    const { value } = this.sortRef.current;
    const { productList } = this.state;
    if (value === 'dsnd') {
      this.setState({
        productList: productList.sort((a, b) => b.id - a.id),
      });
    } else {
      this.setState({
        productList: productList.sort((a, b) => a.id - b.id),
      });
    }
  };

  getValue = () => {
    const { query } = this.props;
    const { value } = this.searchRef.current;
    if (value) {
      query.name = value;
    }
    const queryStr = queryParamStr(query);

    Router.push(`/dashboard?${queryStr}`);
  };

  render() {
    const { query, className } = this.props;
    const { productList } = this.state;
    return (
      <div className={`${className} row`}>
        <div className="col-12">
          <h3>Selected Filters</h3>
          {query &&
            Object.keys(query).map(filter => {
              return this.filterPils(filter);
            })}
        </div>
        <div className="row" style={{ width: '100%' }}>
          <div className="col-9 col-12-small">
            <input
              type="text"
              placeholder="Search by Name"
              ref={this.searchRef}
              className="search-Box"
            />
            <button className="search-btn" onClick={this.getValue}>
              Search
            </button>
          </div>
          <div className="col-3 col-12-small">
            <select className="sort-selector" ref={this.sortRef} onChange={this.getSortedValue}>
              <option value="">Sort By ID</option>
              <option value="asnd">Ascesnding</option>
              <option value="dsnd">Descending</option>
            </select>
          </div>
        </div>
        <div className="row">
          {productList &&
            productList.map(item => {
              return <ProductListGrid key={item.id} product={item} queryOrigin={query.origin} />;
            })}
          {!productList && <div>NO results found.</div>}
        </div>
      </div>
    );
  }
}

ProductListing.propTypes = {
  className: PropTypes.string,
  data: PropTypes.shape([]),
  query: PropTypes.shape({
    name: PropTypes.string,
    origin: PropTypes.string,
  }),
};
ProductListing.defaultProps = {
  className: '',
  data: [],
  query: {
    name: '',
    origin: '',
  },
};

export default withStyles(ProductListing, styles);
export { ProductListing };
