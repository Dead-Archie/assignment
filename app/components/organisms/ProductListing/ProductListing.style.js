import { css } from 'styled-components';

export default css`
  .filter-pills {
    background-color: #ddd;
    border: none;
    color: #868484;
    text-align: center;
    text-decoration: none;
    display: inline-block;
    margin: 4px 2px;
    cursor: pointer;
    border-radius: 16px;
  }
  .filter-pills:after {
    content: ' X';
    padding-left: 10px;
    color: aliceblue;
  }
  .search-Box {
    width: 70%;
    float: left;
  }
`;
