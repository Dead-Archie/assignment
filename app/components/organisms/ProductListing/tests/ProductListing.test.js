import React from 'react';
import { shallow } from 'enzyme';
import 'jest-styled-components';

import { ProductListing } from '../ProductListing';

jest.mock('next/router', () => ({ push: () => {} }));

const sampleProps = {
  className: 'withStyles-sc-4wveuc-0 lgzevM',
  query: {
    name: 'morty',
    origin: 'Earth (C-137)',
  },
  data: [
    {
      name: 'Morty Smith',
      origin: { name: 'Earth (C-137)', url: 'https://rickandmortya' },
      species: 'Human',
      status: 'Alive',
      type: '',
    },
  ],
};

const setUp = props => {
  return shallow(<ProductListing {...props} />);
};

describe('<ProductListing />', () => {
  let ProductListingPageComponent;
  beforeEach(() => {
    ProductListingPageComponent = setUp(sampleProps);
  });

  test('should render with out props correctly', () => {
    const wrapper = shallow(<ProductListing />);
    expect(wrapper).toMatchSnapshot();
  });
  test('should render correctly', () => {
    expect(ProductListingPageComponent).toMatchSnapshot();
  });
  test('check state', () => {
    const wrapper = setUp();
    const checkState = wrapper.state('productList');
    expect(checkState).toEqual([]);
  });
  test('search ref define  ', () => {
    const wrapper = setUp();
    expect(wrapper.instance().searchRef).toEqual({ current: null });
  });
  test('sort ref define  ', () => {
    const wrapper = setUp();
    expect(wrapper.instance().sortRef).toEqual({ current: null });
  });

  test('onclick Button ', () => {
    const wrapper = setUp();
    const searchButton = wrapper.find('.search-btn');
    wrapper.instance().searchRef = {
      current: {
        value: null,
      },
    };
    searchButton.simulate('click');
    expect(wrapper.instance().searchRef).toBeTruthy();
  });

  test('onchange sort Button ', () => {
    const wrapper = setUp();
    const sortButton = wrapper.find('.sort-selector');
    wrapper.instance().sortRef = {
      current: {
        value: null,
      },
    };
    sortButton.simulate('change');
    expect(wrapper.instance().sortRef).toBeTruthy();
  });

  test('if ProductList has values', () => {
    const wrapper = setUp();
    const ProductListGridComp = wrapper.find('ProductListGrid');
    expect(ProductListGridComp).toHaveLength(0);
  });

  test('onclick removePils Button ', () => {
    const wrapper = setUp();
    const removePils = wrapper.find('.filter-pills');
    expect(removePils).toHaveLength(1);
    removePils.simulate('click');
  });
});
