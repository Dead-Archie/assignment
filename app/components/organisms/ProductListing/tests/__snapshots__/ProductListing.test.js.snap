// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductListing /> should render correctly 1`] = `
<div
  className="lgzevM row"
>
  <div
    className="col-12"
  >
    <h3>
      Selected Filters
    </h3>
    <button
      className="filter-pills"
      onClick={[Function]}
    >
      Earth (C-137)
    </button>
  </div>
  <div
    className="row"
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <div
      className="col-9 col-12-small"
    >
      <input
        className="search-Box"
        placeholder="Search by Name"
        type="text"
      />
      <button
        className="search-btn"
        onClick={[Function]}
      >
        Search
      </button>
    </div>
    <div
      className="col-3 col-12-small"
    >
      <select
        className="sort-selector"
        onChange={[Function]}
      >
        <option
          value=""
        >
          Sort By ID
        </option>
        <option
          value="asnd"
        >
          Ascesnding
        </option>
        <option
          value="dsnd"
        >
          Descending
        </option>
      </select>
    </div>
  </div>
  <div
    className="row"
  >
    <withStyles
      product={
        Object {
          "name": "Morty Smith",
          "origin": Object {
            "name": "Earth (C-137)",
            "url": "https://rickandmortya",
          },
          "species": "Human",
          "status": "Alive",
          "type": "",
        }
      }
      queryOrigin="Earth (C-137)"
    />
  </div>
</div>
`;

exports[`<ProductListing /> should render with out props correctly 1`] = `
<div
  className=" row"
>
  <div
    className="col-12"
  >
    <h3>
      Selected Filters
    </h3>
    <button
      className="filter-pills"
      onClick={[Function]}
    />
  </div>
  <div
    className="row"
    style={
      Object {
        "width": "100%",
      }
    }
  >
    <div
      className="col-9 col-12-small"
    >
      <input
        className="search-Box"
        placeholder="Search by Name"
        type="text"
      />
      <button
        className="search-btn"
        onClick={[Function]}
      >
        Search
      </button>
    </div>
    <div
      className="col-3 col-12-small"
    >
      <select
        className="sort-selector"
        onChange={[Function]}
      >
        <option
          value=""
        >
          Sort By ID
        </option>
        <option
          value="asnd"
        >
          Ascesnding
        </option>
        <option
          value="dsnd"
        >
          Descending
        </option>
      </select>
    </div>
  </div>
  <div
    className="row"
  />
</div>
`;
